PackageViewModel/SignPackageViewModel.cs

```diff
@@ -0,0 +1,363 @@
+
+                    foreach (var certificate in store.Certificates)
+                    {
+                        if (CertificateUtility.IsValidForPurposeFast(certificate, Oids.CodeSigningEku))
+                        {
+                            collection.Add(certificate);
+                        }
+                    }
+
+                    var certificates = X509Certificate2UI.SelectFromCollection(
+                        collection,
+                        "Choose a Certificate for Package Signing",
+                        "Provide the code signing certificate for signing the package.",
+                        X509SelectionFlag.SingleSelection);
+
+                    if (certificates.Count > 0)
+                    {
+                        Certificate = certificates[0];
+                        CertificateFileName = null;
+                        ShowPassword = false;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                OnError(ex);
+            }
+        }
+
+        private void ShowCertificateCommandExecute()
+        {
+            var certificate = Certificate;
+            if (certificate != null)
+            {
+                try
+                {
+                    X509Certificate2UI.DisplayCertificate(certificate);
+                }
+                catch { }
+            }
+        }
+
+        // https://github.com/NuGet/NuGet.Client/blob/4c0c9658445573845ddbeff5656e4b3129f727a2/src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandRunner.cs
+        public async Task<string> SignPackage()
+        {
+            ShowProgress = true;
+            CanSign = false;
+            Status = "Signing package...";
+
+            try
+            {
+                // change to AuthorSignPackageRequest when NuGet.Client is updated
+                using (var tempCertificate = new X509Certificate2(Certificate))
+                using (var signRequest = new SignPackageRequest(tempCertificate, GetHashAlgorithmName(tempCertificate)))
+                {
+                    SigningUtility.Verify(signRequest, NullLogger.Instance);
+
+                    var packagePath = _packageViewModel.GetCurrentPackageTempFile();
+                    var originalPackageCopyPath = Path.GetTempFileName();
+
+                    File.Copy(packagePath, originalPackageCopyPath, overwrite: true);
+
+                    using (var packageReadStream = File.OpenRead(packagePath))
+                    using (var packageWriteStream = File.Open(originalPackageCopyPath, FileMode.Open))
+                    using (var package = new SignedPackageArchive(packageReadStream, packageWriteStream))
+                    {
+                        var signer = new Signer(package, new X509SignatureProvider(null));
+                        await Task.Run(() => signer.SignAsync(signRequest, NullLogger.Instance, CancellationToken.None));
```

Comment:
Why do we need a Task.Run here?

Reply:
Otherwise it blocks the UI. I currently don't know why. Looks like it is the same problem as here https://github.com/NuGetPackageExplorer/NuGetPackageExplorer/blob/0fb9b0869bf9f1136000700b87414f687728525b/PackageViewModel/EditablePackageMetadata.cs#L75